File: player.rb
# frozen_string_literal: true
require_relative 'weapon'
require_relative 'shield'
require_relative 'dice'
require_relative 'directions'
module Irrgarten

  class Player
    @@MAX_WEAPONS = 2
    @@MAX_SHIELDS = 3
    @@INITIAL_HEALTH = 10
    @@HITS2LOSE = 3

    def initialize(number, intelligence, strength)
      @name = "Player #" + number.to_s
      @number = number
      @intelligence = intelligence
      @strength = strength
      @health = @@INITIAL_HEALTH
      @row = -1
      @col = -1
      @consecutive_hits = 0

      # #atributos de relaciones
      #se ponen corchetes para crear un array y hay que inicializarlo
      @weapons = []
      @shields = []
    end

    def resurrect
      @weapons.clear
      @shields.clear
      @health = @@INITIAL_HEALTH
      @consecutive_hits = 0
    end

    def row
      @row
    end

    def col
      @col
    end

    def number
      @number
    end

    def pos(row, col)
      @row = row
      @col = col
    end

    def dead
      @health == 0
    end

    def move(direction, valid_moves)
      size = valid_moves.size
      contained = valid_moves.include?(direction)
      if (size > 0) && !contained
        first_element = valid_moves[0]
      else
        direction
      end
    end

    def attack
      sum = 0.0
      sum += @strength
      sum += sum_weapons
    end

    def defend(received_attack)
      manage_hit(received_attack)
    end

    def receive_reward
      rand = Dice.new
      w_reward = rand.weapon_reward
      s_reward = rand.shields_reward

      for i in 1..w_reward
        w_new = new_weapon
        receive_weapon(w_new)
      end

      for i in 1..s_reward
        s_new = new_shield
        receive_shield(s_new)
      end

      rand = Dice.new
      extra_health = rand.health_reward
      @health += extra_health
    end

    def to_s
      "P[" + @name + "," + @number.to_s + "," + @intelligence.to_s + "," + @strength.to_s + "," + @health.to_s + "] \n"
    end

    private

    def receive_weapon(w)
      for i in 0...@@MAX_WEAPONS
        if @weapons.empty?
          wi = @weapons[i]
          discard = wi.discard

          if discard
            @weapons.delete(wi)
          end
        end
      end

      size = @weapons.size

      if size < @@MAX_WEAPONS
        @weapons.push(w)
      end
    end

    def receive_shield(s)
      for i in 0...@@MAX_SHIELDS
        if @shields.empty?
          si = @shields[i]
          discard = si.discard
          if discard
            @shields.delete(si)
          end
        end
      end
      size = @shields.size

      if size < @@MAX_SHIELDS
        @shields.push(s)
      end
    end

    def new_weapon
      rand = Dice.new
      w1 = Weapon.new(rand.weapon_power, rand.uses_left)
      @weapons << w1
    end

    def new_shield
      rand = Dice.new
      s1 = Shield.new(rand.shield_power, rand.uses_left)
      @shields << s1
    end

    def sum_weapons
      sum = 0.0
      if @weapons.empty?
        for weapon in @weapons
          sum += weapon.attack
        end
      end
      sum
    end

    def sum_shields
      sum = 0.0
      if @shields.empty?
        for shield in @shields
          sum += shield.protect.to_f
        end
      end
      sum
    end

    def defensive_energy
      sum = 0.0
      sum += @intelligence
      sum += sum_shields
    end

    def manage_hit(received_attack)
      defense = defensive_energy

      if defense < received_attack
        got_wounded
        inc_consecutive_hits
      else
        reset_hits
      end
      
      if @consecutive_hits == @@HITS2LOSE || dead
        reset_hits
        lose = true
      else
        lose = false
      end

      lose
    end

    def reset_hits
      @consecutive_hits = 0
    end

    def got_wounded
      @health -= 1
    end

    def inc_consecutive_hits
      @consecutive_hits += 1
    end

  end
end

File: Player.java
/**
 *
 * @author gonzaalovd & carlitros_gamer20
 */

package irrgarten;


import java.util.ArrayList;
import irrgarten.Dice;

public class Player extends LabyrinthCharacter{

    static private int MAX_WEAPON = 2;
    static private int MAX_SHIELDS = 3;
    static private int INITIAL_HEALTH = 10;
    static private int HITS2LOSE = 3;

    private char number;
    private int consecutiveHits;
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;

    //atributos de relaciones
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;

    public Player(char number, float intelligence, float strength){

        super("Player #" + number,intelligence,strength, INITIAL_HEALTH);
        this.consecutiveHits = 0;
           
        this.number = number;
        //atribuyos relaciones
        this.weapons = new ArrayList<>();
        this.shields = new ArrayList<>();  
        this.weaponCardDeck = new WeaponCardDeck();
        this.shieldCardDeck = new ShieldCardDeck();
    }
    
    public Player(Player other){
        super(other);
    }

    public void resurrect(){
        this.weapons.clear();
        this.shields.clear();
        super.setHealth(INITIAL_HEALTH);
        this.consecutiveHits = 0;
    }

    public int getRow(){
        return super.getRow();
    }

    public int getCol(){
        return super.getCol();
    }

    public char getNumber(){
        return this.number;

    }

    public void setPos(int row, int col){
        super.setPos(row,col);
    }

    public boolean dead(){
        return super.dead();
    }

    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        /*1.1*/ int size = validMoves.size();
        /*1.2*/ boolean contained = validMoves.contains(direction);
        /*alt*/if(size > 0 && (!contained)){
            /*1.3*/ Directions firstElement = validMoves.get(0);
            /*1.4*/ return firstElement;
        }else{
            /*1.5*/ return direction;
        }
    }

    public float attack(){
        float sum = 0.0f;
        sum += getStrength();
        sum += this.sumWeapons();
        return sum;
    }

    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    } 

    public void recivedReward(){
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for(int i = 0; i < wReward; i++){
            Weapon wnew = weaponCardDeck.nextCard();
            this.receiveWeapon(wnew);
        }
        
        for(int i = 0; i < sReward; i++){
            Shield snew = shieldCardDeck.nextCard();
            this.receiveShield(snew);
        }

        int extraHealth = Dice.healthReward();
        extraHealth += super.getHealth();
        super.setHealth(extraHealth);
        
    }

    public String toString(){
        return "P" + super.toString();
    }

    private void receiveWeapon( Weapon w){
        for(int i = 0; i < this.weapons.size(); i++){
            Weapon wi = this.weapons.get(i);
            boolean discard = wi.discard();
            
            if(discard)
                this.weapons.remove(wi);
        }

        int size = this.weapons.size();
        
        if(size < MAX_WEAPON)
            this.weapons.add(w);
    }

    private void receiveShield(Shield s){
       
        for(int i = 0; i < this.shields.size(); i++){
            Shield si = this.shields.get(i);
            boolean discard = si.discard();
            
            if(discard)
                this.shields.remove(si);
        }

        int size = this.shields.size();
        
        if(size < MAX_SHIELDS)
            this.shields.add(s);
    }

    private Weapon newWeapon(){
        Weapon w1 = new Weapon (Dice.weaponPower(), Dice.usesLeft());

        return w1;
    }

    private Shield newShield(){
        Shield s1 = new Shield (Dice.shieldPower(), Dice.usesLeft());

        return s1;
    }

     protected float sumWeapons(){
        float sum = 0.0f;

        for (Weapon weapon: this.weapons ){
            sum += weapon.attack();
        }
        return sum;
    }

     protected float sumShields(){
         float sum = 0.0f;

         for(Shield shield: this.shields){
             sum += shield.protect();
         }
         return sum;
     }

     protected float defensiveEnergy(){
         float sum= 0.0f;
         sum += super.getIntelligence();
         sum += this.sumShields();
         return sum;
     }

     private boolean manageHit(float receivedAttack){
         float defense = this.defensiveEnergy();

         if(defense < receivedAttack){
             this.gotWounded();
             this.incConsecutiveHits();
         }else{
             this.resetHits();
         }
         
         boolean lose;
         
         if(this.consecutiveHits == HITS2LOSE || this.dead()){
             this.resetHits();
             lose = true;
         }else{
             lose = false;
         }
         
         return lose;
     }

     private void resetHits(){
         this.consecutiveHits = 0;
     }

     private void incConsecutiveHits(){
         this.consecutiveHits++;
     }
}